### PR TITLE
Hash type

### PR DESCRIPTION
**Pull request by pocke**
This pull request adds types to Hash class.
I think we can mark Hash class as "done" by this pull request.

### CODE REVIEW

**Comment by pocke** [comment]
@soutaro It's ready to review now. Could you review this pull request?

File: stdlib/builtin/hash.rbs [hunk 1]
```diff
@@ -1,410 +1,1033 @@
...
+  #     h1[:c]         #=> "c"  # same symbols are all same.
+  #
+  def compare_by_identity: () -> self
+
+  # Returns `true` if *hsh* will compare its keys by their identity.  Also see
+  # Hash#compare_by_identity.
+  #
+  def compare_by_identity?: () -> (true | false)
+
+  def deconstruct_keys: (Array[K]) -> self
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: The parameter can be `nil`?

```
def deconstruct_keys: (Array[K]? keys) -> self
```

File: stdlib/builtin/hash.rbs [hunk 2]
```diff
@@ -1,410 +1,1033 @@
...
-  # see `Hash#compare_by_identity` .
-  def compare_by_identity?: () -> bool
+  # Element Reference---Retrieves the *value* object corresponding to the *key*
+  # object. If not found, returns the default value (see Hash::new for details).
+  #
+  #     h = { "a" => 100, "b" => 200 }
+  #     h["a"]   #=> 100
+  #     h["c"]   #=> nil
+  #
+  def []: (K arg0) -> V?
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: In `Array`, it returns `V` not `V?`. Writing tests anytime we write `[]` seems overkill.

File: stdlib/builtin/hash.rbs [hunk 3]
```diff
@@ -1,410 +1,1033 @@
...
+  #
+  # The second and third form take a single argument which is either an array of
+  # key-value pairs or an object convertible to a hash.
+  #
+  #     Hash["a", 100, "b", 200]             #=> {"a"=>100, "b"=>200}
+  #     Hash[ [ ["a", 100], ["b", 200] ] ]   #=> {"a"=>100, "b"=>200}
+  #     Hash["a" => 100, "b" => 200]         #=> {"a"=>100, "b"=>200}
+  #
+  def self.[]: [U, V] (_ToHash[U, V]) -> ::Hash[U, V]
+             | [U, V] (*[ U, V ]) -> ::Hash[U, V]
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]:
```
| [U, V] (Array[U, V]) -> Hash[U, V]
```

???

File: stdlib/builtin/hash.rbs [hunk 4]
```diff
@@ -1,410 +1,1033 @@
...
+  #     h1.compare_by_identity? #=> true
+  #     h1["a".dup]    #=> nil  # different objects.
+  #     h1[:c]         #=> "c"  # same symbols are all same.
+  #
+  def compare_by_identity: () -> self
+
+  # Returns `true` if *hsh* will compare its keys by their identity.  Also see
+  # Hash#compare_by_identity.
+  #
+  def compare_by_identity?: () -> (true | false)
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: I think `bool` is good for this.

File: stdlib/builtin/hash.rbs [hunk 5]
```diff
@@ -1,410 +1,1033 @@
...
 
-  def keep_if: () { (K arg0, V arg1) -> untyped } -> ::Hash[K, V]
+  # Deletes every key-value pair from *hsh* for which *block* evaluates to
+  # `false`.
+  #
+  # If no block is given, an enumerator is returned instead.
+  #
+  # See also Hash#select!.
+  #
+  def keep_if: () { (K arg0, V arg1) -> untyped } -> self
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]:
```
() { (K, V) -> bool } -> self
```

File: stdlib/builtin/hash.rbs [hunk 6]
```diff
@@ -1,410 +1,1033 @@
...
+  #     h1.merge!(h2, h3) {|key, v1, v2| v1 }
+  #                        #=> {"a"=>100, "b"=>200, "c"=>300, "d"=>400}
+  #     h1                 #=> {"a"=>100, "b"=>200, "c"=>300, "d"=>400}
+  #
+  # Hash#update is an alias for Hash#merge!.
+  #
+  def merge!: [A, B] (*::Hash[A, B] other_hashes) -> ::Hash[A | K, B | V]
+            | [A, B, C] (*::Hash[A, B] other_hashes) { (K key, V oldval, B newval) -> (C) } -> ::Hash[A | K, B | V | C]
+
+  def pretty_print: (untyped pp) -> untyped
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: Could you delete these `pretty_print*` methods?

File: stdlib/builtin/hash.rbs [hunk 7]
```diff
@@ -1,410 +1,1033 @@
...
+
+  # Searches through the hash comparing *obj* with the value using `==`. Returns
+  # the first key-value pair (two-element array) that matches. See also
+  # Array#rassoc.
+  #
+  #     a = {1=> "one", 2 => "two", 3 => "three", "ii" => "two"}
+  #     a.rassoc("two")    #=> [2, "two"]
+  #     a.rassoc("four")   #=> nil
+  #
+  def rassoc: (K arg0) -> [K, V]?
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: `(V) -> [K, V]?` ??

File: stdlib/builtin/hash.rbs [hunk 8]
```diff
@@ -1,410 +1,1033 @@
...
+  #     h = { a: 1, b: 2, c: 3 }
+  #     h.transform_keys! {|k| k.to_s }  #=> { "a" => 1, "b" => 2, "c" => 3 }
+  #     h.transform_keys!(&:to_sym)      #=> { a: 1, b: 2, c: 3 }
+  #     h.transform_keys!.with_index {|k, i| "#{k}.#{i}" }
+  #                                      #=> { "a.0" => 1, "b.1" => 2, "c.2" => 3 }
+  #
+  # If no block is given, an enumerator is returned instead.
+  #
+  def transform_keys!: () -> Enumerator[[ K, V ], Hash[untyped, V]]
+                     | [A] () { (K) -> A } -> Hash[A, V]
```
Comment (soutaro) [inline stdlib/builtin/hash.rbs]: We give monomorphic types for `map!`. How about `() { (K) -> V } -> Hash[K, V]`?

Comment (pocke) [reply stdlib/builtin/hash.rbs]: That makes sense. I updated `transform_keys!` and `transform_values!` typing.
b42db5ae93f30bec36cf6c43b0613017ef1a274a

File: stdlib/builtin/hash.rbs (hunk 3 above)
Comment (pocke) [reply stdlib/builtin/hash.rbs]: Did you mean `[U, V] (Array[[ U, V ]]) -> ::Hash[U, V]`? I fixed with it in 384627caf564dd4051dbde1a3a9fed43a04bf44c.

**Comment by pocke** [comment]
Thanks for your review. I've applied the all reviewed changes.

**Comment by pocke** [comment]
I squashed the commit for review comments to one.
Thanks!

**Pull request merged**